### PR TITLE
`local-file` document connector

### PR DESCRIPTION
**Pull request by zach-blumenfeld**
Adds a deterministic local-file document connector (Markdown / PDF / HTML / AsciiDoc / Word → `Document → Section` graph). After installing the editable build, scaffold a full-stack app **and** ingest a real vault into Neo4j with one command:

```bash
create-context-graph my-analyst-app \
  --domain financial-services --framework pydanticai \
  --connector local-file --local-file-path tests/fixtures/local_file_vault \
  --ingest --neo4j-local
```

No LLMs, no embeddings, no randomness anywhere in the pipeline for MVP.  All connectors need these so adding general modularized components for these later may make more sense.

I am still working on cleaning up documentation test suite design and fixing some markdown formatting issues. 

---

## What's in the branch

Three commits, oldest first.

### 1. `feat(connectors): add local-file document connector` (525d041) — main implementation

The heavy commit. Whole connector + tests + CLI wiring + generated-project template.

- **Data model.** `:Document` and `:Section` nodes keyed on URI in the `name` field — works with the existing `MERGE-on-(name, domain)` pipeline in `ingest.py` with zero changes. Section URI shape: `{document_uri}#{h1-slug}/{h2-slug}/...` using the GitHub/Pandoc slug algorithm (NFKD → ASCII → lowercase → `[a-z0-9_-]` only).
- **Edge cases honored.** Per-parent duplicate-heading disambiguation (`-1`, `-2`); skipped heading levels produce direct H1→H3 children with `headingLevel=3` (no synthetic H2); shallow `description` (body text + `uri:` pointers to direct children only — child body lives on child nodes).
- **`description` field name is deliberate.** Picks up the existing `search_entities` Cypher's `coalesce(n.description, '')` clause for free — no client-side changes needed.
- **Five parsers.**
  - Markdown: `markdown-it-py` GFM tree walk.
  - PDF: three-tier deterministic strategy — `pypdf` outline (covers ~40-60% of long-form PDFs) → `/StructTreeRoot` → `pdfplumber` font-size fallback.
  - HTML: BeautifulSoup + lxml; `` is not a link.
  - AsciiDoc: pure-Python regex parser (no maintained permissive parser exists in 2026); tracks `----` / `....` / `++++` / `|===` block state so `=` inside literal blocks isn't mistaken for a heading.
  - Word `.docx`: `python-docx` style-name detection.
- **Link handling.** Per-format parsers collect raw hrefs only; `DocumentMapper` does classification. `http(s)://` → `URL_LINK` stub `:Document`; relative paths resolved against the source doc's directory (deterministic textual resolution, never touches the filesystem); `mailto:` / `tel:` / `javascript:` / `data:` / `ftp:` / `sms:` skipped. Stub upsert semantics let order-independent re-ingest converge to the same graph.
- **Cross-platform URIs.** Every Document URI is `Path(p).resolve().as_posix()` — same file produces the same URI on macOS, Linux, and Windows. File discovery is POSIX-sorted, so output ordering is deterministic.
- **CLI.** Five `--local-file-*` options (`--path` repeatable, `--pattern`, `--recursive` / `--no-recursive`, `--follow-links`, `--exclude` repeatable). Passing `--connector local-file` without a path errors loudly rather than silently importing CWD.
- **Wizard.** Deliberate spec deviation: `get_credential_prompts()` returns **one required path prompt** rather than `[]`, so a wizard user can't accidentally pick the connector and ingest random files.
- **Generated-project template.** `templates/backend/connectors/local_file_connector.py.j2` is a self-contained markdown-only copy in `{% raw %}` — same pattern as `claude_code_connector.py.j2`. Dispatch table is structured so adding the other parsers downstream is a one-case change.
- **Dependencies.** Per existing convention, the seven new libraries (`markdown-it-py`, `mdit-py-plugins`, `pypdf`, `pdfplumber`, `beautifulsoup4`, `lxml`, `python-docx`) are appended to the shared `connectors` extra rather than a new `local-file` group. All MIT/BSD-3 — verified no AGPL/GPL.
- **Tests.** +72 (1,113 → 1,185 passing). 95-test connector suite covers slug edge cases, POSIX URI normalization, every per-format parser (reportlab-generated PDFs gated behind `importorskip`), DocumentMapper link classification / dedup / stub-upgrade, and the connector class itself. CLI tests cover the required-path error, dry-run, template generation, and full flag matrix.

### 2. `test(local-file): add opt-in functional test over realistic vault fixture` (e9041f5) — end-to-end vault

The 95-test unit suite covers one behavior per test with hand-crafted inputs. This adds a companion suite that ingests an entire realistic Obsidian-style notebook through `LocalFileConnector.fetch()` and asserts structural invariants — catches integration regressions no individual unit test would (parser tweaks that produce orphan sections under cross-doc anchors, slug changes that create duplicate-key Documents under a full vault, etc.).

- **Fixture: `tests/fixtures/local_file_vault/`** — fictional financial analyst's notebook for ACME Incorporated's Corporate Strategy & FP&A team. 17 files exercising every parser path: 11× Markdown (vault README, daily notes, company analyses, people atomic notes, meeting capture, decision trace, methodology), 1× AsciiDoc with `[source]----…----` listing blocks, 1× HTML with hierarchical anchors + mailto footer, 1× 4-page PDF with a 3-level outline, 1× DOCX with Heading 1/2/3 styles. Inline `[[wikilinks]]` throughout (Obsidian-style; must remain plain text, not become LINKS_TO edges).
- **Meta-doc placement.** `tests/fixtures/local_file_vault_TESTING.md` lives *next to* the vault, not inside — putting it in the vault would mean the connector ingests it on every run and pollutes the graph.
- **Opt-in via `--functional` pytest flag** (mirrors `--integration`, `--slow`). Default `make test` clean-skips all 35 tests; `make test-functional` runs them in ~0.9s.
- **18 edge cases asserted** across 7 test classes — full table in the commit message.

### 3. `docs(local-file): document end-to-end app run against the vault fixture` (050cd80) — running the app

Expands the in-fixture TESTING meta-doc with a "Running the connector / app against this vault" section answering the practical question: *how do I go from these committed fixtures to a running graph-backed app?*

- **Option A — in-memory smoke test** (no Neo4j, no scaffold): the same `LocalFileConnector.fetch()` call the functional test makes, as a one-liner for fast parser iteration.
- **Option B — full scaffold + ingest:** pick a Neo4j (Desktop / `neo4j-local` / Aura), install the editable build with `[connectors,dev]`, then run the headline command above (three variants shown: explicit creds, `--neo4j-local`, `--neo4j-aura-env`).
- **`make install && make dev`** to start the FastAPI + Next.js app, plus a short list of demo prompts the agent answers well on this vault.
- **Cypher snippets** for inspecting the resulting graph in Neo4j Browser, including the lazy-loading "unfetched URL stubs" query from spec §6.5.
- **`--local-file-*` flag combinations** covering markdown-only ingest, multi-root + exclude, and single-file ingestion.

Consolidates the practical content from the original `scratch/test-vault-cli-guide.md` into a committed location that survives `scratch/` being gitignored.

---

## Test plan

- [x] `make test` — 1,185 passing, 263 skipped (35 new vault-test skips on top of the pre-existing 228); same 15 pre-existing `hr-technical-resume` failures unrelated to this change
- [x] `make test-functional` — 35 passed in ~0.9s
- [x] `make lint` — ruff clean across `src/` and `tests/`
- [ ] Reviewer: run the headline `create-context-graph ... --ingest --neo4j-local` command and confirm a graph appears in Neo4j Browser
- [ ] Reviewer: spot-check the Obsidian frontmatter behavior (markdown-it-py treats `---` after text as a Setext H2 — see TESTING.md §6 for whether this needs a follow-up `mdit-py-plugins.front_matter` fix or is acceptable for MVP)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/create_context_graph/connectors/local_file_connector.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            if not root.exists():
+                logger.warning("Path does not exist, skipping: %s", root)
+                continue
+
+            if root.is_file():
+                candidates: list[Path] = [root]
+            elif self._recursive:
+                candidates = list(root.rglob(self._pattern))
+            else:
+                candidates = list(root.glob(self._pattern))
```
[the remark this answers is not in the document]
Comment (jexp) [reply src/create_context_graph/connectors/local_file_connector.py:196]: @copilot apply changes based on reject ** if recursive is false [this feedback](https://github.com/neo4j-labs/create-context-graph/pull/34#discussion_r3222534141)

File: src/create_context_graph/templates/backend/connectors/local_file_connector.py.j2 [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+        for root in self._paths:
+            if not root.exists():
+                logger.warning("Path does not exist, skipping: %s", root)
+                continue
+            if root.is_file():
+                candidates = [root]
+            elif self._recursive:
+                candidates = list(root.rglob(self._pattern))
+            else:
+                candidates = list(root.glob(self._pattern))
```
[the remark this answers is not in the document]
Comment (jexp) [reply src/create_context_graph/templates/backend/connectors/local_file_connector.py.j2:535]: @copilot apply changes based on [this feedback](https://github.com/neo4j-labs/create-context-graph/pull/34#discussion_r3222534179) reject `**` if recursive is false

File: src/create_context_graph/connectors/_local_file/mapper.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    def __init__(self) -> None:
+        # Per-entity dedup: URI → entity dict. Insertion order preserved.
+        self._documents: dict[str, dict] = {}
+        self._sections: dict[str, dict] = {}
+        # All relationships emitted so far.
+        self._relationships: list[dict] = []
+        # URIs of documents we parsed during this run (used to distinguish
+        # "real" from "stub" link targets at the end).
+        self._parsed_doc_uris: set[str] = set()
+        # Stamp once per run — the existing ingest pipeline overwrites
```
[the remark this answers is not in the document]
Comment (jexp) [reply src/create_context_graph/connectors/_local_file/mapper.py:94]: @copilot fix it so that it matches the docstring behaivor to drive stub/upsert [this feedback](https://github.com/neo4j-labs/create-context-graph/pull/34#discussion_r3222534200)

**Comment by jexp** [comment]
## PDF Parsing Library Research: Optional PyMuPDF Tier 0

After reviewing all major Python PDF parsing libraries, here's where things stand:

### Library Comparison

| Library | License | Speed (complex 72p PDF) | Heading support | Notes |
|---------|---------|------------------------|-----------------|-------|
| **PyMuPDF** (`fitz`) | **AGPL-3.0** | **~0.33s** | TOC + font heuristic | 5k+ stars, Artifex-backed, most capable |
| pypdf | BSD-3 | ~1.8s | Outline + StructTree | Already bundled (tiers 1–2) |
| pdfplumber | MIT | ~6.6s | Font heuristic | Already bundled (tier 3) |
| pdfminer.six | MIT | ~5s | Font heuristic | pdfplumber wraps it; no advantage to switching |
| pypdfium2 | Apache-2/BSD-3 | Fast raw | None | Raw text only, no heading/layout analysis |
| **PDF Oxide** | **MIT/Apache-2** | **0.8ms** (text extraction) | `detect_headings=True` | Newer (732 stars, v0.3.46, 53 releases) |

### Decision: Optional PyMuPDF Tier 0

PyMuPDF is **~10× faster than pypdf** and **~20× faster than pdfplumber** on complex documents, but its AGPL-3.0 license means we cannot bundle it. Solution: make it a **user-installable opt-in** that is detected at runtime:

```bash
pip install PyMuPDF
# or in a generated project:
uv add PyMuPDF
```

When present, it runs as **tier 0** before the existing pypdf/pdfplumber tiers. No configuration needed. This keeps the default distribution fully BSD/MIT while giving high-volume users a significant performance boost.

**Tier 0 has two sub-paths:**
- **0a (TOC)**: Uses `get_toc()` — same signal as tier 1 (PDF outline) but with faster page text extraction
- **0b (font heuristic)**: Uses `get_text("dict")` block API — same most-frequent-size algorithm as tier 3 but ~20× faster than pdfplumber's char-by-char iteration

### What about PDF Oxide?

[PDF Oxide](https://pdf.oxide.fyi) (MIT/Apache-2, `pip install pdf-oxide`) is an exciting option:
- **5.8× faster than PyMuPDF** for raw text extraction (0.8ms vs 4.6ms mean on 3,830 PDFs, 100% pass rate)
- Supports `detect_headings=True` for structured Markdown output  
- v0.3.46, 53 releases, 22 contributors, most recent release May 11 2026

However it's newer and smaller ecosystem than PyMuPDF's established tooling. I've noted it in the docs as a planned tier 0b — tracking it for a future PR once heading extraction is more battle-tested.

### Changes in this commit

- `pdf.py`: `_try_pymupdf()` with TOC path (`_pymupdf_sections_from_toc`) and font heuristic path (`_pymupdf_font_heuristic`), plus URI link extraction (`_pymupdf_uri_links`). `parse()` calls it first as tier 0.
- `TestPDFParserPyMuPDF` (4 tests): mocked `fitz` — no PyMuPDF install needed in CI
- `docs/docs/how-to/import-local-files.md`: new page covering formats, CLI options, graph shape, 4-tier PDF performance table, PyMuPDF AGPL note, PDF Oxide mention
